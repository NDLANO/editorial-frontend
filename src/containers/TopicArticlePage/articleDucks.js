/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { handleActions, createAction } from 'redux-actions';

import { createSelector } from 'reselect';
import { getLocale } from '../Locale/localeSelectors';
import { titleI18N, tagsI18N, introductionI18N, metaDescriptionI18N, contentI18N } from '../../util/i18nFieldFinder';
import formatDate from '../../util/formatDate';

export const fetchArticle = createAction('FETCH_ARTICLE');
export const setArticle = createAction('SET_ARTICLE');
export const updateArticle = createAction('UPDATE_ARTICLE');

export const actions = {
  updateArticle,
  fetchArticle,
  setArticle,
};

const initalState = {};

export default handleActions({
  [setArticle]: {
    next: (state, action) => ({ ...state, [action.payload.id]: { ...action.payload } }),
    throw: state => state,
  },
}, initalState);

const getArticleFromState = state => state.articles;

export const getArticleById = articleId => createSelector(
  [getArticleFromState],
  articles => articles[articleId],
);

export const getArticle = articleId => createSelector(
  [getArticleById(articleId), getLocale],
  (article, locale) => (
    article ? {
      ...article,
      title: titleI18N(article, locale, true),
      introduction: introductionI18N(article, locale, true),
      content: contentI18N(article, locale, true),
      metaDescription: metaDescriptionI18N(article, locale, true),
      tags: tagsI18N(article, locale, true),
      created: formatDate(article.created, locale),
      updated: formatDate(article.updated, locale),
    } : undefined
  ),
);
