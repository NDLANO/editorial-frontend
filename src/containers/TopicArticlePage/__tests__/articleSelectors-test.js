/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { getArticle } from '../articleDucks';

const state = {
  locale: 'nb',
  articles: {
    all: {
      1: {
        id: 1,
        created: '2014-12-24T10:44:06Z',
        title: [
            { title: 'Tester', language: 'nb' },
            { title: 'Testing', language: 'en' },
        ],
        metaDescription: [
        { metaDescription: 'Beskrivelse', language: 'nb' },
        { metaDescription: 'Description', language: 'en' },
        ],
      },
      2: {
        id: 2,
        created: '2014-11-24T10:44:06Z',
        title: [
          { title: 'Tester', language: 'nb' },
          { title: 'Testing', language: 'en' },
        ],
      },
    },
  },
};

test('articleSelectors getArticle with id', () => {
  expect(getArticle(1)(state).id).toBe(1);
  expect(getArticle(2)(state).id).toBe(2);
});

test('articleSelectors getArticle nb locale', () => {
  const getArticleSelector = getArticle(1);
  expect(getArticleSelector(state).id).toBe(1);
  expect(getArticleSelector(state).title).toBe('Tester');
  expect(getArticleSelector(state).created).toBe('24.12.2014');
  expect(getArticleSelector(state).metaDescription).toBe('Beskrivelse');
});

test('articleSelectors getArticle en locale', () => {
  const getArticleSelector = getArticle(1);
  const stateWithEnLocale = { ...state, locale: 'en' };
  expect(getArticleSelector(stateWithEnLocale).id).toBe(1);
  expect(getArticleSelector(stateWithEnLocale).title).toBe('Testing');
  expect(getArticleSelector(stateWithEnLocale).created).toBe('12/24/2014');
  expect(getArticleSelector(stateWithEnLocale).metaDescription).toBe('Description');
});

test('articleSelectors getArticle returns undefined if article is not in state', () => {
  const getArticleSelector = getArticle(3);
  expect(getArticleSelector(state)).toEqual(undefined);
});
