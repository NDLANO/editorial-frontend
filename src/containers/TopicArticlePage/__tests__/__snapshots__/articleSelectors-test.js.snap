// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`articleSelectors getArticle (en locale) 1`] = `
Object {
  "articleType": "topic-article",
  "content": "<section><p>Tim Berners Lees the World Wide Web is...</p></section>",
  "copyright": Object {
    "authors": Array [
      Object {
        "name": "Ragna Marie TÃ¸rdal",
        "type": "Forfatter",
      },
    ],
    "license": Object {
      "description": "Creative Commons Attribution-ShareAlike 2.0 Generic",
      "license": "by-sa",
      "url": "https://creativecommons.org/licenses/by-sa/2.0/",
    },
    "origin": "",
  },
  "created": "11/26/2016",
  "id": "1",
  "introduction": "Brainstorming is an important",
  "metaDescription": "Brainstorming",
  "oldNdlaUrl": "//ndla.no/node/172732",
  "requiredLibraries": Array [],
  "revision": 1,
  "tags": Array [
    "storytelling",
    "idea development",
    "media design",
    "instruments",
  ],
  "title": "Brainstorming",
  "updated": "12/01/2016",
  "updatedBy": "content-import-client",
  "visualElement": "<embed/>",
}
`;

exports[`articleSelectors getArticle (nb locale) 1`] = `[Function]`;
